File: sdks/php/src/Results.php
<?php

declare(strict_types=1);

namespace MongoDo;

/**
 * InsertOneResult - Result of an insertOne operation.
 */
readonly class InsertOneResult
{
    public function __construct(
        public string|int $insertedId,
        public bool $acknowledged = true,
    ) {}

    /**
     * Get the inserted document ID.
     */
    public function getInsertedId(): string|int
    {
        return $this->insertedId;
    }

    /**
     * Check if the write was acknowledged.
     */
    public function isAcknowledged(): bool
    {
        return $this->acknowledged;
    }
}

/**
 * InsertManyResult - Result of an insertMany operation.
 */
readonly class InsertManyResult
{
    /**
     * @param array<int, string|int> $insertedIds
     */
    public function __construct(
        public array $insertedIds,
        public int $insertedCount,
        public bool $acknowledged = true,
    ) {}

    /**
     * Get the inserted document IDs.
     *
     * @return array<int, string|int>
     */
    public function getInsertedIds(): array
    {
        return $this->insertedIds;
    }

    /**
     * Get the count of inserted documents.
     */
    public function getInsertedCount(): int
    {
        return $this->insertedCount;
    }

    /**
     * Check if the write was acknowledged.
     */
    public function isAcknowledged(): bool
    {
        return $this->acknowledged;
    }
}

/**
 * UpdateResult - Result of an update/replace operation.
 */
readonly class UpdateResult
{
    public function __construct(
        public int $matchedCount = 0,
        public int $modifiedCount = 0,
        public string|int|null $upsertedId = null,
        public bool $acknowledged = true,
    ) {}

    /**
     * Get the count of matched documents.
     */
    public function getMatchedCount(): int
    {
        return $this->matchedCount;
    }

    /**
     * Get the count of modified documents.
     */
    public function getModifiedCount(): int
    {
        return $this->modifiedCount;
    }

    /**
     * Get the upserted document ID, if any.
     */
    public function getUpsertedId(): string|int|null
    {
        return $this->upsertedId;
    }

    /**
     * Check if the write was acknowledged.
     */
    public function isAcknowledged(): bool
    {
        return $this->acknowledged;
    }

    /**
     * Check if a document was upserted.
     */
    public function wasUpserted(): bool
    {
        return $this->upsertedId !== null;
    }
}

/**
 * DeleteResult - Result of a delete operation.
 */
readonly class DeleteResult
{
    public function __construct(
        public int $deletedCount = 0,
        public bool $acknowledged = true,
    ) {}

    /**
     * Get the count of deleted documents.
     */
    public function getDeletedCount(): int
    {
        return $this->deletedCount;
    }

    /**
     * Check if the write was acknowledged.
     */
    public function isAcknowledged(): bool
    {
        return $this->acknowledged;
    }
}

/**
 * BulkWriteResult - Result of a bulk write operation.
 */
readonly class BulkWriteResult
{
    /**
     * @param array<int, string|int> $upsertedIds
     */
    public function __construct(
        public int $insertedCount = 0,
        public int $matchedCount = 0,
        public int $modifiedCount = 0,
        public int $deletedCount = 0,
        public int $upsertedCount = 0,
        public array $upsertedIds = [],
        public bool $acknowledged = true,
    ) {}

    /**
     * Get the count of inserted documents.
     */
    public function getInsertedCount(): int
    {
        return $this->insertedCount;
    }

    /**
     * Get the count of matched documents.
     */
    public function getMatchedCount(): int
    {
        return $this->matchedCount;
    }

    /**
     * Get the count of modified documents.
     */
    public function getModifiedCount(): int
    {
        return $this->modifiedCount;
    }

    /**
     * Get the count of deleted documents.
     */
    public function getDeletedCount(): int
    {
        return $this->deletedCount;
    }

    /**
     * Get the count of upserted documents.
     */
    public function getUpsertedCount(): int
    {
        return $this->upsertedCount;
    }

    /**
     * Get the upserted document IDs.
     *
     * @return array<int, string|int>
     */
    public function getUpsertedIds(): array
    {
        return $this->upsertedIds;
    }

    /**
     * Check if the write was acknowledged.
     */
    public function isAcknowledged(): bool
    {
        return $this->acknowledged;
    }
}

File: sdks/kotlin/src/main/kotlin/do_/mongo/Results.kt
package do_.mongo

/**
 * Result of an insertOne operation.
 */
data class InsertOneResult(
    val acknowledged: Boolean,
    val insertedId: Any?
) {
    companion object {
        @Suppress("UNCHECKED_CAST")
        fun fromDocument(result: Any?): InsertOneResult {
            return when (result) {
                is Document -> InsertOneResult(
                    acknowledged = result.getBoolean("acknowledged", true),
                    insertedId = result["insertedId"]
                )
                is Map<*, *> -> InsertOneResult(
                    acknowledged = (result["acknowledged"] as? Boolean) ?: true,
                    insertedId = result["insertedId"]
                )
                else -> InsertOneResult(acknowledged = true, insertedId = null)
            }
        }
    }
}

/**
 * Result of an insertMany operation.
 */
data class InsertManyResult(
    val acknowledged: Boolean,
    val insertedIds: Map<Int, Any?>
) {
    val insertedCount: Int get() = insertedIds.size

    companion object {
        @Suppress("UNCHECKED_CAST")
        fun fromDocument(result: Any?): InsertManyResult {
            return when (result) {
                is Document -> InsertManyResult(
                    acknowledged = result.getBoolean("acknowledged", true),
                    insertedIds = parseInsertedIds(result["insertedIds"])
                )
                is Map<*, *> -> InsertManyResult(
                    acknowledged = (result["acknowledged"] as? Boolean) ?: true,
                    insertedIds = parseInsertedIds(result["insertedIds"])
                )
                else -> InsertManyResult(acknowledged = true, insertedIds = emptyMap())
            }
        }

        @Suppress("UNCHECKED_CAST")
        private fun parseInsertedIds(ids: Any?): Map<Int, Any?> {
            return when (ids) {
                is Map<*, *> -> ids.entries.associate { (k, v) ->
                    (k.toString().toIntOrNull() ?: 0) to v
                }
                is List<*> -> ids.mapIndexed { index, id -> index to id }.toMap()
                else -> emptyMap()
            }
        }
    }
}

/**
 * Result of an update operation.
 */
data class UpdateResult(
    val acknowledged: Boolean,
    val matchedCount: Long,
    val modifiedCount: Long,
    val upsertedId: Any?
) {
    val wasAcknowledged: Boolean get() = acknowledged
    val upsertedCount: Long get() = if (upsertedId != null) 1 else 0

    companion object {
        @Suppress("UNCHECKED_CAST")
        fun fromDocument(result: Any?): UpdateResult {
            return when (result) {
                is Document -> UpdateResult(
                    acknowledged = result.getBoolean("acknowledged", true),
                    matchedCount = result.getLong("matchedCount") ?: 0,
                    modifiedCount = result.getLong("modifiedCount") ?: 0,
                    upsertedId = result["upsertedId"]
                )
                is Map<*, *> -> UpdateResult(
                    acknowledged = (result["acknowledged"] as? Boolean) ?: true,
                    matchedCount = (result["matchedCount"] as? Number)?.toLong() ?: 0,
                    modifiedCount = (result["modifiedCount"] as? Number)?.toLong() ?: 0,
                    upsertedId = result["upsertedId"]
                )
                else -> UpdateResult(acknowledged = true, matchedCount = 0, modifiedCount = 0, upsertedId = null)
            }
        }
    }
}

/**
 * Result of a delete operation.
 */
data class DeleteResult(
    val acknowledged: Boolean,
    val deletedCount: Long
) {
    val wasAcknowledged: Boolean get() = acknowledged

    companion object {
        @Suppress("UNCHECKED_CAST")
        fun fromDocument(result: Any?): DeleteResult {
            return when (result) {
                is Document -> DeleteResult(
                    acknowledged = result.getBoolean("acknowledged", true),
                    deletedCount = result.getLong("deletedCount") ?: 0
                )
                is Map<*, *> -> DeleteResult(
                    acknowledged = (result["acknowledged"] as? Boolean) ?: true,
                    deletedCount = (result["deletedCount"] as? Number)?.toLong() ?: 0
                )
                else -> DeleteResult(acknowledged = true, deletedCount = 0)
            }
        }
    }
}

/**
 * Result of a bulk write operation.
 */
data class BulkWriteResult(
    val acknowledged: Boolean,
    val insertedCount: Int,
    val matchedCount: Int,
    val modifiedCount: Int,
    val deletedCount: Int,
    val upsertedCount: Int,
    val insertedIds: Map<Int, Any?>,
    val upsertedIds: Map<Int, Any?>
) {
    companion object {
        @Suppress("UNCHECKED_CAST")
        fun fromDocument(result: Any?): BulkWriteResult {
            return when (result) {
                is Document -> BulkWriteResult(
                    acknowledged = result.getBoolean("acknowledged", true),
                    insertedCount = result.getInt("insertedCount") ?: 0,
                    matchedCount = result.getInt("matchedCount") ?: 0,
                    modifiedCount = result.getInt("modifiedCount") ?: 0,
                    deletedCount = result.getInt("deletedCount") ?: 0,
                    upsertedCount = result.getInt("upsertedCount") ?: 0,
                    insertedIds = (result["insertedIds"] as? Map<Int, Any?>) ?: emptyMap(),
                    upsertedIds = (result["upsertedIds"] as? Map<Int, Any?>) ?: emptyMap()
                )
                else -> BulkWriteResult(
                    acknowledged = true,
                    insertedCount = 0,
                    matchedCount = 0,
                    modifiedCount = 0,
                    deletedCount = 0,
                    upsertedCount = 0,
                    insertedIds = emptyMap(),
                    upsertedIds = emptyMap()
                )
            }
        }
    }
}
